Auth: Allow getting a TLS permission checker when filtering used-by URLs

markylaing opened the pull request:
Adds a way to circumvent the default behaviour of the TLS authorizer which returns 403 Forbidden when getting a permission checker. 

Note that I'm not 100% convinced by this solution, passing things around in the context is useful but in this case it is modifying the behaviour of the authorizer. It feels overly obfuscated. That said, there is precedent for it already in the use of `request.CtxEffectiveProjectName`.

Please see the linked issue and commit messages for more detailed information.

Closes #13533.

tomponline commented:
> An exception to this was added for the project entity type to
enable filtering when listing projects. A further exception needs
to be made so that used-by URLs can be filtered correctly. This
can now be accomplished using the new context key.

What is the exception needed for filtering used-by urls (do these not project param in them?)

markylaing commented:
> > An exception to this was added for the project entity type to
> > enable filtering when listing projects. A further exception needs
> > to be made so that used-by URLs can be filtered correctly. This
> > can now be accomplished using the new context key.
> 
> What is the exception needed for filtering used-by urls (do these not project param in them?)

`GetPermissionChecker` needs to behave differently in a few scenarios. Let's say we have a TLS identity that is restricted, and only has access to `example-project`.
1. `GET /1.0/instances?project=example-project`: Allow all instance URLs that have `example-project` as the project.
2. `GET /1.0/instances`: Return 403 Forbidden - default project is implicit.
3. `GET /1.0/projects`: Filter out all projects except for `example-project`. Ignore that the implicit project for the request is `default`.
4. `GET /1.0/projects?recursion=1`: As in 3, then when filtering the used-by URLs also ignore that the implicit project for the request is `default`.

When I say "implicit" project above. I strictly mean the result of `request.ProjectParam`. The TLS authorizer was using it quite blindly to enforce cases 1 and 2, it handled case 3 directly, but did not handle the fourth case.

tomponline commented:
> > > An exception to this was added for the project entity type to
> > > enable filtering when listing projects. A further exception needs
> > > to be made so that used-by URLs can be filtered correctly. This
> > > can now be accomplished using the new context key.
> > 
> > 
> > What is the exception needed for filtering used-by urls (do these not project param in them?)
> 
> `GetPermissionChecker` needs to behave differently in a few scenarios. Let's say we have a TLS identity that is restricted, and only has access to `example-project`.
> 
>     1. `GET /1.0/instances?project=example-project`: Allow all instance URLs that have `example-project` as the project.
> 
>     2. `GET /1.0/instances`: Return 403 Forbidden - default project is implicit.
> 
>     3. `GET /1.0/projects`: Filter out all projects except for `example-project`. Ignore that the implicit project for the request is `default`.
> 
>     4. `GET /1.0/projects?recursion=1`: As in 3, then when filtering the used-by URLs also ignore that the implicit project for the request is `default`.
> 
> 
> When I say "implicit" project above. I strictly mean the result of `request.ProjectParam`. The TLS authorizer was using it quite blindly to enforce cases 1 and 2, it handled case 3 directly, but did not handle the fourth case.

Can the permission checker function be extended to take a (list?) of allowed projects? Or does that push too much of the auth logic out of the authorizer where its supposed to stay?

markylaing commented:
> > When I say "implicit" project above. I strictly mean the result of `request.ProjectParam`. The TLS authorizer was using it quite blindly to enforce cases 1 and 2, it handled case 3 directly, but did not handle the fourth case.
> 
> Can the permission checker function be extended to take a (list?) of allowed projects? Or does that push too much of the auth logic out of the authorizer where its supposed to stay?

This is what I found tricky on this PR. If we drop [this block](https://github.com/canonical/lxd/blob/81000f8c89235e2fd88100715900f2fb7d7d8230/lxd/auth/driver_tls.go#L197-L200) from the TLS authorizer we would have to perform a separate auth check from `GetPermissionChecker`. We could add a boolean to ignore the check in certain scenarios.

Actually I think I've just had an idea that might be a better solution. I'll update the PR and see what you think.

markylaing commented:
My idea was to create a separate access handler for this. But this has repeated lots of logic from the authorizers so it's not as good a solution as I thought.

While working on it, it has occurred to me that case 2 (from https://github.com/canonical/lxd/pull/13539#issuecomment-2147452241) will no longer apply when fine-grained authorization is allowed for TLS identities. This is because it will no longer be possible to use the project name as a basis for what an identity is allowed to view (e.g. the instances they can view will be based entirely on `can_view` for each instance and resolved by the auth model). This will constitute a change in API behaviour for TLS users but I don't see a way around it if we want fine-grained auth to function properly.

markylaing commented:
@tomponline I've updated the PR with what we discussed yesterday and added to the `tls_restrictions` test such that currently fails on edge and passes on this branch. I'll update the PR description now too.

tomponline commented:
Thanks!
